backends/bmv2/jsonconverter.cpp

```diff
@@ -1348,25 +1476,73 @@ cstring JsonConverter::convertHashAlgorithm(cstring algorithm) const {
     return result;
 }
 
+/**
+ Computes the type of the key based on the declaration in the path.
+
+If the key expression invokes .isValid(), we return "valid".
+@param ke A table match key element
+@return the key type
+*/
+cstring JsonConverter::getKeyMatchType(const IR::KeyElement *ke)
+{
+    auto mt = refMap->getDeclaration(ke->matchType->path, true)->to<IR::Declaration_ID>();
+    BUG_CHECK(mt != nullptr, "%1%: could not find declaration", ke->matchType);
+    auto expr = ke->expression;
+
+    cstring match_type = mt->name.name;
+    if (mt->name.name == corelib.exactMatch.name || mt->name.name == corelib.ternaryMatch.name) {
+        if (expr->is<IR::MethodCallExpression>()) {
+            auto mi = P4::MethodInstance::resolve(expr->to<IR::MethodCallExpression>(),
+                                                  refMap, typeMap);
+            if (mi->is<P4::BuiltInMethod>())
+                if (mi->to<P4::BuiltInMethod>()->name == IR::Type_Header::isValid)
+                    match_type = "valid";
+        }
+    } else if (mt->name.name != corelib.lpmMatch.name &&
+               mt->name.name != v1model.rangeMatchType.name &&
+               mt->name.name != v1model.selectorMatchType.name) {
+        ::error("%1%: match type not supported on this target", mt);
+    }
+    return match_type;
+}
+
 Util::IJson*
 JsonConverter::convertTable(const CFG::TableNode* node,
                             Util::JsonArray* counters,
-                            Util::JsonArray* action_profiles) {
+                            Util::JsonArray* action_profiles)
+{
     auto table = node->table;
     LOG3("Processing " << dbp(table));
     auto result = new Util::JsonObject();
     cstring name = table->externalName();
     result->emplace("name", name);
     result->emplace("id", nextId("tables"));
-    cstring table_match_type = "exact";
+    cstring table_match_type = corelib.exactMatch.name;
```

Comment:
What @mbudiu-vmw is saying is that I could very well have chosen to use the string "EXACT" instead of "exact" in the JSON, and that it is a "coincidence" that I am using the same string value.